### PR TITLE
Fix/#7/prod ready

### LINKED ISSUE #7: assignment2.1
Create an issue for the current assignment and fill it out as discussed in the previous module on GIT practices.

Create a branch off of the tag v0.2 for the changes requested in this exercise and name it appropriately (again, as discussed in the previous module on GIT practices).

Try to apply as much of the principles discussed in the guidelines document and your own prior experience to reformat and refactor the provided script into a version that is production ready and create a PR.

The submitted version should have the following:

An updated README  document on how to install/run your code

An env.yaml which is the conda environment file that records all the required dependencies of your script. The README should talk about how to install your code and run it.

Make your code packageable and installable as a library ([Here is a reference ](https://python-packaging-tutorial.readthedocs.io/en/latest/setup_py.html)to python packaging). Refer to the section on the recommended project layout.

Basic tests (can be scripts) to verify correct installation. Again, refer to the section on project layout for where to place these tests.

Create ML workflow scripts:
A script (ingest_data.py)to download and create training and validation datasets. The script should accept the output folder/file path as an user argument.
A script (train.py) to train the model(s). The script should accept arguments for input (dataset) and output folders (model pickles).
A script (score.py) to score the model(s). The script should accept arguments for model folder, dataset folder and any outputs.

NOTES:
For the above scripts, use the argparse module to accept user arguments when running the scripts. Here is a [quick tutorial ](https://docs.python.org/3/howto/argparse.html)on how to use argparse.
Ensure that your scripts generate a log for each script run ([Here is Tiger’s reference ](https://tigeranalytics-code-templates.readthedocs-hosted.com/en/latest/developer_guide/coding_best_practices.html#logging)with some example code on logging). Support accepting arguments (to your scripts) for the following
specify the log level (e.g. `--log-level DEBUG`)
use a log file or not (e.g. `--log-path <path>`, default case would be to not write logs to a file)
toggle whether or not to write logs to the console (e.g. `--no-console-log`, default case would be to always write logs to console)
When arguments are not specified, use the appropriate folders from the recommended project layout as the defaults (e.g. data, artifacts, logs etc.)
Feel free to use a configuration file for the default arguments. But, the scripts should still override the config file values if explicitly passed by the user.

Add unit tests and functional tests to your library ([Here is a short introduction ](https://tigeranalytics-code-templates.readthedocs-hosted.com/en/latest/developer_guide/coding_best_practices.html#code-testing)to testing principles and some example code).

Use Sphinx to generate HTML documents for your library ([Here is a short write up ](https://tigeranalytics-code-templates.readthedocs-hosted.com/en/latest/developer_guide/coding_best_practices.html#documentation)on Sphinx and an example). Note that, at Tiger, we use the [Numpy docstring style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html). Pls. ensure the docstrings follow the same style.

### PR DESCRIPTION
**Pull request by deepak-sharma-ta**
fixes #7

### CODE REVIEW

File: src/ingest_data.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import sys
+
+sys.path.append("/home/deepak/practice_git_repo/assignment2/mle-training/")
+from src.log_config import LOGGING_DEFAULT_CONFIG, configure_logger
+
+
+def fetch_housing_data(housing_url, housing_path):
+    """Download the housing data
+
+    :param housing_url: Url of the dataset
```
Comment (vinayksoni) [inline src/ingest_data.py]: read instruction - Tiger uses nympy docstring formatting

File: src/ingest_data.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+    housing_tgz = tarfile.open(tgz_path)
+    housing_tgz.extractall(path=housing_path)
+    housing_tgz.close()
+
+
+def load_housing_data(housing_path):
+    """Download the housing data
+
+    :param housing_path: Path where the dataset is stored
```
Comment (vinayksoni) [inline src/ingest_data.py]: read instruction - Tiger uses nympy docstring formatting

File: src/log_config.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+
+def configure_logger(
+    logger=None, cfg=None, log_file=None, console=True, log_level="DEBUG"
+):
+    """Function to setup configurations of logger through function.
+
+    The individual arguments of `log_file`, `console`, `log_level` will overwrite the ones in cfg.
+
```
Comment (vinayksoni) [inline src/log_config.py]: read instruction - Tiger uses nympy docstring formatting

File: src/ingest_data.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    r = requests.get(housing_url, timeout=300)
+    with open(tgz_path, "wb") as f:
+        f.write(r.content)
+
+    housing_tgz = tarfile.open(tgz_path)
+    housing_tgz.extractall(path=housing_path)
+    housing_tgz.close()
+
+
+def load_housing_data(housing_path):
```
Comment (vinayksoni) [inline src/ingest_data.py]: add datatype

File: src/log_config.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        },
+        "simple": {"format": "ARTHUR %(message)s"},
+    },
+    "root": {"level": "DEBUG"},
+    "loggers": {},
+}
+
+
+def configure_logger(
+    logger=None, cfg=None, log_file=None, console=True, log_level="DEBUG"
```
Comment (vinayksoni) [inline src/log_config.py]: add datatype

File: src/score.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+import numpy as np
+import pandas as pd
+import sys
+from sklearn.metrics import mean_absolute_error, mean_squared_error
+
+sys.path.insert(0, os.path.abspath("."))
+from src.log_config import LOGGING_DEFAULT_CONFIG, configure_logger
+
+
+def lin_reg_pred(MODEL_PATH, housing_prepared, housing_labels):
```
Comment (vinayksoni) [inline src/score.py]: add datatype

File: src/score.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    """Make model prediction
+
+    :param MODEL_PATH: Path of the model
+    :param housing_prepared: Pre-processed housing dataset with features
```
Comment (vinayksoni) [inline src/score.py]: read instruction - Tiger uses nympy docstring formatting

File: src/score.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    lin_mae = mean_absolute_error(housing_labels, housing_predictions)
+
+    return lin_rmse, lin_mae
+
+
+def dtree_pred(MODEL_PATH, housing_prepared, housing_labels):
+    """Make model prediction
+
+    :param MODEL_PATH: Path of the model
+    :param housing_prepared: Pre-processed housing dataset with features
```
Comment (vinayksoni) [inline src/score.py]: read instruction - Tiger uses nympy docstring formatting

File: src/score.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    housing_predictions = LIN_REG.predict(housing_prepared)
+
+    lin_mse = mean_squared_error(housing_labels, housing_predictions)
+    lin_rmse = np.sqrt(lin_mse)
+    lin_mae = mean_absolute_error(housing_labels, housing_predictions)
+
+    return lin_rmse, lin_mae
+
+
+def dtree_pred(MODEL_PATH, housing_prepared, housing_labels):
```
Comment (vinayksoni) [inline src/score.py]: add datatype

File: src/score.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+    tree_mse = mean_squared_error(housing_labels, housing_predictions)
+    tree_rmse = np.sqrt(tree_mse)
+    return tree_mse, tree_rmse
+
+
+def rforest_pred(MODEL_PATH, X_test_prepared, y_test):
+    """Make model prediction
+
+    :param MODEL_PATH: Path of the model
```
Comment (vinayksoni) [inline src/score.py]: read instruction - Tiger uses nympy docstring formatting

File: src/score.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    """
+    D_TREE = pickle.load(open(os.path.join(MODEL_PATH, "d_tree.pkl"), "rb"))
+    housing_predictions = D_TREE.predict(housing_prepared)
+
+    tree_mse = mean_squared_error(housing_labels, housing_predictions)
+    tree_rmse = np.sqrt(tree_mse)
+    return tree_mse, tree_rmse
+
+
+def rforest_pred(MODEL_PATH, X_test_prepared, y_test):
```
Comment (vinayksoni) [inline src/score.py]: add datatype

File: src/train.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+from sklearn.tree import DecisionTreeRegressor
+
+sys.path.insert(0, os.path.abspath("."))
+from src.log_config import LOGGING_DEFAULT_CONFIG, configure_logger
+
+
+def income_cat_proportions(data):
+    """Find the income proportion per category
+
+    :param data: Path of the daaset
```
Comment (vinayksoni) [inline src/train.py]: read instruction - Tiger uses nympy docstring formatting

File: src/train.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+from sklearn.impute import SimpleImputer
+from sklearn.linear_model import LinearRegression
+from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
+from sklearn.tree import DecisionTreeRegressor
+
+sys.path.insert(0, os.path.abspath("."))
+from src.log_config import LOGGING_DEFAULT_CONFIG, configure_logger
+
+
+def income_cat_proportions(data):
```
Comment (vinayksoni) [inline src/train.py]: add datatype

File: tests/functional_tests/test_training.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+import pandas as pd
+from scipy.stats import randint
+from sklearn.ensemble import RandomForestRegressor
+from sklearn.impute import SimpleImputer
+from sklearn.linear_model import LinearRegression
+from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
+from sklearn.tree import DecisionTreeRegressor
+
+
+PATH = "/home/deepak/practice_git_repo/assignment2/mle-training/data/processed"
```
Comment (vinayksoni) [inline tests/functional_tests/test_training.py]: for pytest cases you should now read from paths. all data has to be pasted as fixture.

File: tests/functional_tests/test_training.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    "bedrooms_per_room",
+    "population_per_household",
+    "ocean_proximity_INLAND",
+    "ocean_proximity_ISLAND",
+    "ocean_proximity_NEAR BAY",
+    "ocean_proximity_NEAR OCEAN",
+]
+
+
+def test_lin_reg_train(
```
Comment (vinayksoni) [inline tests/functional_tests/test_training.py]: for pytest cases you should now read/write from paths.

File: tests/unit_tests/test_data_ingestion.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+import pandas as pd
+import numpy as np
+
+sys.path.append("/home/deepak/practice_git_repo/assignment2/mle-training/")
+
+from src.ingest_data import (
+    load_housing_data,
+)
+
+MODEL_PATH = "/home/deepak/practice_git_repo/assignment2/mle-training/artifacts/"
```
Comment (vinayksoni) [inline tests/unit_tests/test_data_ingestion.py]: for pytest cases you should now read/write from paths.

File: setup.py
Comment (vinayksoni) [inline setup.py:1]: it should not be empty for packaging this

File: artifacts/d_tree.pkl
Comment (vinayksoni) [inline artifacts/d_tree.pkl:1]: git push should not have model files/data. cross check with instruction

**Review by vinayksoni — CHANGES_REQUESTED** [review]

File: src/housing_package/score.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+def lin_reg_pred(
+    MODEL_PATH: str, housing_prepared: pd.DataFrame, housing_labels: pd.DataFrame
+):
+    """Make model prediction
+    Note: Model should exist in the given path
+
+    Parameters
+    ----------
+
+    param1 : str
```
Comment (vinayksoni) [inline src/housing_package/score.py]: Parameters
------------
MODEL_PATH : str
______path of the model

Comment (vinayksoni) [reply src/housing_package/score.py]: do this for all doc strs

**Review by vinayksoni — CHANGES_REQUESTED** [review]

File: src/housing_package/score.py (hunk 17 above)
Comment (deepak-sharma-ta) [reply src/housing_package/score.py]: done

File: src/housing_package/score.py [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+    MODEL_PATH : str
+        Path of the model
+    housing_prepared : csv
+        Pre-processed housing dataset with features
+    housing_labels : csv
+        Target labels of the housing dataset
+
+    Returns
+    -----------
+    rmse, mae
```
Comment (vinayksoni) [inline src/housing_package/score.py]: variable name : datatye
description
vatiable name : datatpe 2
description

File: src/housing_package/score.py [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+
+    MODEL_PATH : str
+        Path of the model
+    housing_prepared : csv
+        Pre-processed housing dataset with features
+    housing_labels : csv
+        Target labels of the housing dataset
+
+    Returns
+    -----------
```
Comment (vinayksoni) [inline src/housing_package/score.py:71]: variable name : datatye
description
vatiable name : datatpe 2
description

File: src/housing_package/score.py [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+
+    MODEL_PATH : str
+        Path of the model
+    X_test_prepared : csv
+        Pre-processed housing dataset with features
+    y_test : csv
+        Target labels of the housing dataset
+
+    Returns
+    -----------
```
Comment (vinayksoni) [inline src/housing_package/score.py:102]: variable name : datatye
description
vatiable name : datatpe 2
description

File: src/housing_package/train.py [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+
+    Parameters
+    ----------
+
+    data : str
+        Path of the dataset
+
+    Returns
+    -----------
+    proportion of income category
```
Comment (vinayksoni) [inline src/housing_package/train.py]: variable name : datatye
description
vatiable name : datatpe 2
description

File: tests/unit_tests/test_data_ingestion.py [hunk 22]
```diff
@@ -0,0 +1,61 @@
+import pickle
+import sys
+import os
+import pandas as pd
+import numpy as np
+import requests
+import tarfile
+
+sys.path.append("/home/deepak/practice_git_repo/assignment2/mle-training/")
```
Comment (vinayksoni) [inline tests/unit_tests/test_data_ingestion.py]: py test should not deal with any paths.... once you package this and zip it .... it goes across different systems and different platforms.... and their installation of your packages will fail.

Comment (vinayksoni) [reply tests/unit_tests/test_data_ingestion.py]: you can copy and create data set (can be  a subset) here and then train your model and then get the results and verify the pipeline

**Review by vinayksoni — CHANGES_REQUESTED** [review]

**Review by vinayksoni — APPROVED** [review]

**Pull request merged**